pygmt/src/subplot.py

```diff
@@ -0,0 +1,200 @@
+            lib.call_module(module="subplot", args=arg_str)
+            # Setup matplotlib-like Axes
+            axs = np.empty(shape=(nrows, ncols), dtype=object)
+            for index in range(nrows * ncols):
+                i = index // ncols  # row
+                j = index % ncols  # column
+                axs[i, j] = index
+            yield axs
+        finally:
+            v_arg = build_arg_string(kwargs.fromkeys("V"))
+            lib.call_module("subplot", f"end {v_arg}".strip())
+
+
+@fmt_docstring
+@contextlib.contextmanager
+@use_alias(A="fixedlabel", C="clearance", V="verbose")
+def sca(self, ax=None, **kwargs):
+    r"""
+    Set the current Axes instance to *ax*.
+
+    Before you start plotting you must first select the active subplot. Note:
+    If any *projection* option is passed with **?** as scale or width when
+    plotting subplots, then the dimensions of the map are automatically
+    determined by the subplot size and your region. For Cartesian plots: If you
+    want the scale to apply equally to both dimensions then you must specify
+    ``projection="x"`` [The default ``projection="X"`` will fill the subplot by
+    using unequal scales].
+
+    {aliases}
+
+    Parameters
+    ----------
+    ax : str
+        *row,col*\|\ *index*.
+        Sets the current subplot until further notice.  **Note**: First *row*
+        or *col* is 0, not 1. If not given we go to the next subplot by order
+        specified via **autolabel**.  As an alternative, you may bypass the
+        **sca** mode and instead supply the common option **ax**\ [*row,col*]
+        to the first plot command you issue in that subplot. GMT maintains
+        information about the current figure and subplot. Also, you may give
+        the one-dimensional *index* instead which starts at 0 and follows the
+        row or column order set via **autolabel**.
+
+    fixedlabel : str
+        Overrides the automatic labeling with the given string. No modifiers
+        are allowed. Placement, justification, etc. are all inherited from how
+        **autolabel** was specified by the initial **subplot** command.
+
+    clearance : str
+        [*side*]\ *clearance*.
+        Reserve a space of dimension *clearance* between the margin and the
+        subplot on the specified side, using *side* values from **w**, **e**,
+        **s**, or **n**.  The option is repeatable to set aside space on more
+        than one side.  Such space will be left untouched by the main map
+        plotting but can be accessed by modules that plot scales, bars, text,
+        etc.  This setting overrides the common clearances set by **clearance**
+        in the initial **subplot** call.
+
+    {V}
+    """
+    kwargs = self._preprocess(**kwargs)  # pylint: disable=protected-access
+
+    with Session() as lib:
+        arg_str = " ".join(["set", f"{ax}", build_arg_string(kwargs)])
+        lib.call_module(module="subplot", args=arg_str)
+        yield
```

Comment:
> > @weiji14 Any thought to using `sca` in a context manager? I know it doesn't have a `sca_end` equivalent the way that `inset` and `subplot` do, but I think using it with `with` statements will make it more clear that the code is creating different subplots, as opposed to a long line of similar looking plotting functions.
> 
> I did think of this, but it's technically quite hard to implement. What should a `with fig.sca()` block reset to? The first subplot? The next subplot? There might be a way using function decorators (?) to insert `ax=1` for all `fig.*` functions under the with block but I'll need more time to confirm that it's possible. Or feel free to suggest ideas :)

`fig.sca` is now a context manager as implemented here. There's not `try... finally` block because there's nothing to reset to. Have a look at the unit tests and tutorials and let me know what you think (i.e. if `with fig.sca():` is a good syntax to have, or if we should just stick to `fig.sca()`).

Reply:
Yes, the `fig.sca` context manager makes no sense here. I'm wondering if we can ask for upstream changes to be able to reset the ax number?

Reply:
But that would have to wait until GMT 6.2.0 :slightly_smiling_face: Also as mentioned at https://github.com/GenericMappingTools/pygmt/pull/822#issuecomment-770345781, what do we reset to? The first subplot? The previous one (-1)?

Reply:
> The first subplot? The previous one (-1)?

Both doesn't make sense to me. It's weird and confusing to have `sca` as a context manager, but can add plotting elements to panels outside the context manager.

Reply:
Yeah, but map elements outside a `with fig.sca()` block can't just go into the main figure, as it would need to be outside the `fig.subplot()` indent. It's hard to come up with a good solution here.

So should we 1) just go back to having `fig.sca` as an ordinary function (i.e. **not** a context manager)? Or 2) keep `fig.sca()` as a context manager so that people can organize their subplots using indentation (if they choose to, or not)?